### PR TITLE
RM-26

### PR DESCRIPTION
**Pull request by Alinadimmokdad**
new page iv-item-parts
table grid no delete no add just edit on edit gridToolbar to add item parts

### CODE REVIEW

**Comment by omarZrayka23** [comment]
Screenshot 2024-11-05 at 9 04 39 AM

i believes first column is mandatory

File: src/pages/iv-item-parts/forms/ItemPartForm.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+      label: labels.qty,
+      name: 'qty',
+      props: {
+        allowNegative: false,
+        decimalScale: 0
+      }
+    }
+  ]
+
+  function getData() {
```
Comment (omarZrayka23) [inline src/pages/iv-item-parts/forms/ItemPartForm.js:125]: If you are certain that you do not need getData somewhere in your code, you can use useEffect directly and insert the method getData in useEffect.

**Comment by imtithalhalalo** [comment]
Screenshot 2024-11-05 155741
page not added

File: src/pages/iv-item-parts/forms/ItemPartForm.js [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+        decimalScale: 0
+      }
+    }
+  ]
+
+  useEffect(() => {
+    getRequest({
+      extension: InventoryRepository.ItemParts.qry,
+      parameters: `_itemId=${obj.recordId}`
+    }).then(res => {
```
Comment (imtithalhalalo) [inline src/pages/iv-item-parts/forms/ItemPartForm.js:136]: remove .then catch

File: src/repositories/InventoryRepository.js [hunk 3]
```diff
@@ -204,5 +204,15 @@ export const InventoryRepository = {
...
+  },
+  IVMDParts: {
+    qry: service + 'qryIT'
+  },
+  ItemParts: {
+    qry: service + 'qryITP',
+    set2: service + 'set2ITP'
+  },
+  Parts: {
+    snapshot: service + 'snapshotPRT'
```
Comment (imtithalhalalo) [inline src/repositories/InventoryRepository.js:216]: is this snapshot used? if not please remove it

File: src/pages/iv-item-parts/forms/ItemPartForm.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+      if (items.length === 1 && isRowEmpty(items[0])) {
+        postData({ itemId: values.itemId, items: [] })
+      } else {
+        postData(values)
+      }
+    }
+  })
+
+  const isRowEmpty = row => {
```
Comment (imtithalhalalo) [inline src/pages/iv-item-parts/forms/ItemPartForm.js:64]: add this function at the top of formik

File: src/pages/iv-item-parts/forms/ItemPartForm.js [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    enableReinitialize: true,
+    validateOnChange: true,
+
+    validationSchema: yup.object({
+      items: yup
+        .array()
+        .of(
+          yup.object().shape({
+            qty: yup.number().test(function (value) {
+              return isRowEmpty(this.parent) || value > 0
```
Comment (imtithalhalalo) [inline src/pages/iv-item-parts/forms/ItemPartForm.js:30]: for value > 0 you cannot use allowNegative = false?

File: src/pages/iv-item-parts/forms/ItemPartForm.js [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+        allowNegative: false,
+        decimalScale: 0
+      }
+    },
+    {
+      component: 'numberfield',
+      label: labels.qty,
+      name: 'qty',
+      props: {
+        allowNegative: false,
```
Comment (imtithalhalalo) [inline src/pages/iv-item-parts/forms/ItemPartForm.js:126]: then why do you need it in validation? value > 0

File: src/pages/iv-item-parts/forms/ItemPartForm.js [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        ]
+      }
+    },
+    {
+      component: 'textfield',
+      label: labels.partName,
+      name: 'partName',
+      props: {
+        readOnly: true,
+        allowNegative: false,
```
Comment (imtithalhalalo) [inline src/pages/iv-item-parts/forms/ItemPartForm.js:117]: if it is textField why do you need allowNegative and decimalScale?

**Comment by omarZrayka23** [comment]
[
Screenshot 2024-11-07 at 8 28 25 AM
](url)

When I clicked submit after clearing the section, everything was successful, however row one was missing.

**Review by omarZrayka23 — APPROVED** [review]

**Comment by NoraBitarSM** [comment]
i could add 2 empty rows 
image

**Pull request merged**